File: C-Source/pub1/Distrib/LIB/LIB-ForBIN/TestDownload/CSH/StartProcess.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NIT.Test.Download._45._000
{
    public class StartProcess
    {
        // Set a Private Members
        private String strCmdPath;
        private String strCmd32Path;
        private String strCScriptPath;
        private String strWScriptPath;
        private String strWPoshPath;
        private String strCScript32Path;
        private String strWPosh32Path;
        private String strMSIExecPath;
        private String strWMICPath;
        private String strNotepadPath;

        // Private Methods
        private void SetPaths()
        {
            strCmdPath = UserDirectories.GetPathCmd() + "\\cmd.exe";
            if(!File.Exists(strCmdPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File {0} is not found.", strCmdPath);
                strCmdPath = String.Empty;
            }
            strCmd32Path = UserDirectories.GetPathWOWCmd() + "\\cmd.exe";
            if (!File.Exists(strCmd32Path))
            {
                if(UserDirectories.Is64Bit())
                    Console.Error.WriteLine("Check Integrity Error. The File {0} is not found.", strCmd32Path);
                strCmd32Path = String.Empty;
            }
            strCScriptPath = UserDirectories.GetPathCmd() + "\\cscript.exe";
            if (!File.Exists(strCScriptPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File {0} is not found.", strCScriptPath);
                strCScriptPath = String.Empty;
            }
            strWScriptPath = UserDirectories.GetPathCmd() + "\\wscript.exe";
            if (!File.Exists(strWScriptPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File {0} is not found.", strWScriptPath);
                strWScriptPath = String.Empty;
            }
            strWPoshPath = UserDirectories.GetPathCmd() + "\\WindowsPowerShell\\v1.0\\powershell.exe";
            if (!File.Exists(strWPoshPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File {0} is not found.", strWPoshPath);
                strWPoshPath = String.Empty;
            }
            strCScript32Path = UserDirectories.GetCScipt32Path();
            if (!File.Exists(strCScript32Path))
            {
                Console.Error.WriteLine("Check Integrity Error. The File CScript 32 {0} is not found.", strCScript32Path);
                strCScript32Path = String.Empty;
            }
            strWPosh32Path = UserDirectories.GetPoshEXE32Path();
            if (!File.Exists(strWPosh32Path))
            {
                Console.Error.WriteLine("Check Integrity Error. The File PowerSghell 32 {0} is not found.", strWPosh32Path);
                strWPosh32Path = String.Empty;
            }
            strMSIExecPath = UserDirectories.GetMSIEXECPath();
            if (!File.Exists(strMSIExecPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File MSIExec 32 {0} is not found.", strMSIExecPath);
                strMSIExecPath = String.Empty;
            }
            strWMICPath = UserDirectories.GetWMICEXEPath();
            if (!File.Exists(strWMICPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File WMIC 32 {0} is not found.", strWMICPath);
                strWMICPath = String.Empty;
            }

            strNotepadPath = UserDirectories.GetSystemRootPath() + "\\notepad.exe";
            if (!File.Exists(strNotepadPath))
            {
                Console.Error.WriteLine("Check Integrity Error. The File {0} is not found.", strNotepadPath);
                strNotepadPath = String.Empty;
            }
        }

        // Constructor

        public StartProcess()
        {
            SetPaths();
        }

        // Public Methods

        public int StartAsFile(String strFile, int iTimeOut)
        {
            if (strFile.Length == 0)
            {
                Console.Error.WriteLine("Process Error! Can\'t Run Empty File.");
                return 1;
            }
            if (!File.Exists(strFile))
            {
                Console.Error.WriteLine("Error open file: {0}.\nFile may contains a malware or does not exist.\n", strFile);
                return 1;
            }
            ProcessStartInfo pri = new ProcessStartInfo();
            Process firstProc = new Process();
            pri.FileName = strFile;
            firstProc.EnableRaisingEvents = true;
            firstProc.StartInfo = pri;

            try
            {
                firstProc.Start();
                if (iTimeOut <= 0)
                {
                    firstProc.WaitForExit();
                    Console.Out.WriteLine("Process {0} is terminated Successfully!");
                    return 0;
                }
                else
                {
                    if (!firstProc.WaitForExit(iTimeOut))
                    {
                        firstProc.Kill();
                    }
                    Console.Out.WriteLine("Process {0} is terminated after {1} milliseconds.", strFile, iTimeOut);
                    return 0;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Process Error. Can\'t Run a File {0}.\n{1}", strFile, e.Message);
                return 2;
            }
        }

        public int StartAsNotepad(String strFile, int iTimeOut)
        {
            if (strFile.Length == 0)
            {
                Console.Error.WriteLine("Process Error! Can\'t Run Empty File.");
                return 1;
            }
            if (!File.Exists(strFile))
            {
                Console.Error.WriteLine("Error open file: {0}.\nFile may contains a malware or does not exist.\n", strFile);
                return 1;
            }
            if (strNotepadPath.Length == 0)
            {
                Console.Error.WriteLine("Check Integrity Error! The Notepad Path {0} does not Exists!", strNotepadPath);
                return 1;
            }
            ProcessStartInfo pri = new ProcessStartInfo();
            Process firstProc = new Process();
            pri.FileName = strNotepadPath;
            pri.Arguments = strFile;
            pri.RedirectStandardOutput = false;
            pri.UseShellExecute = true;
            pri.CreateNoWindow = false;
            firstProc.EnableRaisingEvents = true;
            firstProc.StartInfo = pri;

            try
            {
                firstProc.Start();
                if (iTimeOut <= 0)
                {
                    firstProc.WaitForExit();
                    Console.Out.WriteLine("Process {0} is terminated Successfully!");
                    return 0;
                }
                else
                {
                    if (!firstProc.WaitForExit(iTimeOut))
                    {
                        firstProc.Kill();
                    }
                    Console.Out.WriteLine("Process {0} is terminated after {1} milliseconds.", strFile, iTimeOut);
                    return 0;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Process Error. Can\'t Run a File {0}.\n{1}", strFile, e.Message);
                return 2;
            }
        }

        public int StartAsCmdFile(String strFile, int iTimeOut)
        {
            String strArgs;
            // Console.Out.WriteLine("File {0}", strFile);
            if (strFile.Length == 0)
            {
                Console.Error.WriteLine("Process Error! Can\'t Run Empty File.");
                return 1;
            }
            if(!File.Exists(strFile))
            {
                Console.Error.WriteLine("Error open file: {0}.\nFile may contains a malware or does not exist.\n", strFile);
                return 1;
            }
            if (strCmdPath.Length == 0)
            {
                Console.Error.WriteLine("Check Integrity Error! The Cmd Path {0} does not Exists!", strCmdPath);
                return 1;
            }
            ProcessStartInfo pri = new ProcessStartInfo();
            Process firstProc = new Process();
            pri.FileName = strCmdPath;
            strArgs = " /c \"" + strFile + "\"";
            pri.Arguments = strArgs;
            pri.RedirectStandardOutput = false;
            pri.UseShellExecute = true;
            pri.CreateNoWindow = false;
            firstProc.EnableRaisingEvents = true;
            firstProc.StartInfo = pri;

            try
            {
                firstProc.Start();
                if (iTimeOut <= 0)
                {
                    firstProc.WaitForExit();
                    Console.Out.WriteLine("Process {0} is terminated Successfully!");
                    return 0;
                }
                else
                {
                    if (!firstProc.WaitForExit(iTimeOut))
                    {
                        firstProc.Kill();
                    }
                    Console.Out.WriteLine("Process {0} is terminated after {1} milliseconds.", strFile, iTimeOut);
                    return 0;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Process Error. Can\'t Run a File {0}.\n{1}", strFile, e.Message);
                return 2;
            }
        }

        public int StartAsPoshFile(String strFile, int iTimeOut)
        {
            String strArgs;
            if (strFile.Length == 0)
            {
                Console.Error.WriteLine("Process Error! Can\'t Run Empty File.");
                return 1;
            }
            if (!File.Exists(strFile))
            {
                Console.Error.WriteLine("Error open file: {0}.\nFile may contains a malware or does not exist.\n", strFile);
                return 1;
            }
            if (strWPoshPath.Length == 0)
            {
                Console.Error.WriteLine("Check Integrity Error! The Powershell Path {0} does not Exists!", strWPoshPath);
                return 1;
            }
            ProcessStartInfo pri = new ProcessStartInfo();
            Process firstProc = new Process();
            pri.FileName = strWPoshPath;
            strArgs = " -NoLogo -NoProfile -ExecutionPolicy Bypass -File \"" + strFile + "\"";
            pri.Arguments = strArgs;
            pri.RedirectStandardOutput = false;
            pri.UseShellExecute = true;
            pri.CreateNoWindow = false;
            firstProc.EnableRaisingEvents = true;
            firstProc.StartInfo = pri;

            try
            {
                firstProc.Start();
                if (iTimeOut <= 0)
                {
                    firstProc.WaitForExit();
                    Console.Out.WriteLine("Process {0} is terminated Successfully!");
                    return 0;
                }
                else
                {
                    if (!firstProc.WaitForExit(iTimeOut))
                    {
                        firstProc.Kill();
                    }
                    Console.Out.WriteLine("Process {0} is terminated after {1} milliseconds.", strFile, iTimeOut);
                    return 0;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Process Error. Can\'t Run a File {0}.\n{1}", strFile, e.Message);
                return 2;
            }
        }

        public int StartAsCScriptFile(String strFile, int iTimeOut)
        {
            String strArgs;
            if (strFile.Length == 0)
            {
                Console.Error.WriteLine("Process Error! Can\'t Run Empty File.");
                return 1;
            }
            if (!File.Exists(strFile))
            {
                Console.Error.WriteLine("Error open file: {0}.\nFile may contains a malware or does not exist.\n", strFile);
                return 1;
            }
            if (strCScriptPath.Length == 0)
            {
                Console.Error.WriteLine("Check Integrity Error! The CScript Path {0} does not Exists!", strCScriptPath);
                return 1;
            }
            ProcessStartInfo pri = new ProcessStartInfo();
            Process firstProc = new Process();
            pri.FileName = strCScriptPath;
            strArgs = " //NoLogo \"" + strFile + "\"";
            pri.Arguments = strArgs;
            pri.RedirectStandardOutput = false;
            pri.UseShellExecute = true;
            pri.CreateNoWindow = false;
            firstProc.EnableRaisingEvents = true;
            firstProc.StartInfo = pri;

            try
            {
                firstProc.Start();
                if (iTimeOut <= 0)
                {
                    firstProc.WaitForExit();
                    Console.Out.WriteLine("Process {0} is terminated Successfully!");
                    return 0;
                }
                else
                {
                    if (!firstProc.WaitForExit(iTimeOut))
                    {
                        firstProc.Kill();
                    }
                    Console.Out.WriteLine("Process {0} is terminated after {1} milliseconds.", strFile, iTimeOut);
                    return 0;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Process Error. Can\'t Run a File {0}.\n{1}", strFile, e.Message);
                return 2;
            }
        }

        public int StartFile(String strFile, int iTimeOut)
        {
            String strArgs;
            if (strFile.Length == 0)
            {
                Console.Error.WriteLine("Process Error! Can\'t Run Empty File.");
                return 1;
            }
            if (!File.Exists(strFile))
            {
                Console.Error.WriteLine("Error open file: {0}.\nFile may contains a malware or does not exist.\n", strFile);
                return 1;
            }
            if (strCmdPath.Length == 0)
            {
                Console.Error.WriteLine("Check Integrity Error! The Cmd Path {0} does not Exists!", strCmdPath);
                return 1;
            }

            ProcessStartInfo pri = new ProcessStartInfo();
            Process firstProc = new Process();
            pri.FileName = strCmdPath;
            strArgs = " /c START /WAIT /REALTIME /B " + strFile;
            pri.Arguments = strArgs;
            pri.RedirectStandardOutput = false;
            pri.UseShellExecute = true;
            pri.CreateNoWindow = false;
            firstProc.EnableRaisingEvents = true;
            firstProc.StartInfo = pri;

            try
            {
                firstProc.Start();
                if (iTimeOut <= 0)
                {
                    firstProc.WaitForExit();
                    Console.Out.WriteLine("Process {0} is terminated Successfully!");
                    return 0;
                }
                else
                {
                    if (!firstProc.WaitForExit(iTimeOut))
                    {
                        firstProc.Kill();
                    }
                    Console.Out.WriteLine("Process {0} is terminated after {1} milliseconds.", strFile, iTimeOut);
                    return 0;
                }
            }
            catch (Exception e)
            {
                Console.Error.WriteLine("Process Error. Can\'t Run a File {0}.\n{1}", strFile, e.Message);
                return 2;
            }
        }

    }

}

File: C-Source/pub1/Distrib/LIB/LIB-ForBIN/TestDownload/CPP/StartProcess.cpp
#include "StartProcess.h"
#include <Windows.h>
#include <iostream>
#include <processthreadsapi.h>
#include <synchapi.h>
#include "UserDirectories.h"

void StartProcess::SetPaths()
{
	UserDirectories stdVar;
	strCmdPath = stdVar.GetCMDEXEPath();
	strCmd32Path = stdVar.GetCMD32EXEPath();
	strWPoshPath = stdVar.GetPOSHEXEPath();
	strCscriptPath = stdVar.GetCSCRIPTEXEPath();
	strCscript32Path = stdVar.GetCSCRIPT32EXEPath();
	strWPosh32Path = stdVar.GetPOSH32EXEPath();
	strMSIExecPath = stdVar.GetMSIEXECPath();
	strWMICPath = stdVar.GetWMIEXEPath();
	strNotepadPath = stdVar.GetNOTEPADEXEPath();
}

int StartProcess::StartAsFile(string strFile, int iTimeOut) {
	STARTUPINFOA stInfo;
	PROCESS_INFORMATION pri;

	DWORD dwsi_size = sizeof(stInfo);
	ZeroMemory(&stInfo, dwsi_size);
	stInfo.cb = dwsi_size;
	ZeroMemory(&pri, sizeof(pri));

	if (strFile.length() == 0) {
		std::cerr << "Process Error! Can\'t Run Empty File.\n";
		return 1;
	}

	DWORD dwTimeOut;
	try
	{
		dwTimeOut = iTimeOut * 1000;
	}
	catch (const std::exception&)
	{
		dwTimeOut = 0;
	}

	// Start Child Process
	if (!CreateProcessA(NULL,
		(LPSTR)strFile.c_str(),
		NULL,
		NULL,
		false,
		IDLE_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS,
		NULL,
		NULL,
		&stInfo,
		&pri
	)) {
		std::cerr << "Process Creation Failed (" << GetLastError() << ").\n";
		return 1;
	}
	else
	{
		if (dwTimeOut == 0) {
			WaitForSingleObject(pri.hProcess, INFINITE);
			std::cout << "Process " << strFile << " is Terminated Successfully!\n";
		}
		else
		{
			WaitForSingleObject(pri.hProcess, dwTimeOut);
			std::cout << "Process " << strFile << " is Terminated after " << dwTimeOut << " milliseconds!\n";
		}

		if (!TerminateProcess(pri.hProcess, 0))
			std:cerr << "Process " << GetProcessId(pri.hProcess) << " is not Terminated!\n";
		// Close Process and Thread handles
		CloseHandle(pri.hProcess);
		CloseHandle(pri.hThread);
	}
	return 0;
}

int StartProcess::StartAsNotepad(string strFile, int iTimeOut) {
	STARTUPINFOA stInfo;
	PROCESS_INFORMATION pri;
	DWORD dwTimeOut;
	string strCmdLine;
	DWORD dwResult;
	int iRes;

	DWORD dwsi_size = sizeof(stInfo);
	ZeroMemory(&stInfo, dwsi_size);
	stInfo.cb = dwsi_size;
	ZeroMemory(&pri, sizeof(pri));

	if (strFile.length() == 0) {
		std::cerr << "Process Error! Can\'t Run Empty File.\n";
		return 1;
	}
	if (strNotepadPath.length() == 0) {
		std::cerr << "Check Integrity Error! The Notepad Path " << strNotepadPath << " does not exist!\n";
		return 1;
	}

	strCmdLine = strNotepadPath + " \"" + strFile + "\"";

	try
	{
		dwTimeOut = iTimeOut * 1000;
	}
	catch (const std::exception&)
	{
		dwTimeOut = 0;
	}

	// Start Child Process
// Start Child Process
	if (!CreateProcessA(NULL,
		(LPSTR)strCmdLine.c_str(),
		NULL,
		NULL,
		false,
		IDLE_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS,
		NULL,
		NULL,
		&stInfo,
		&pri
	)) {
		std::cerr << "Process Creation Failed (" << GetLastError() << ").\n";
		return 1;
	}
	else
	{
		if (dwTimeOut == 0) {
			WaitForSingleObject(pri.hProcess, INFINITE);
			std::cout << "Process " << strCmdLine << " is Terminated Successfully!\n";
		}
		else
		{
			WaitForSingleObject(pri.hProcess, dwTimeOut);
			std::cout << "Process " << strCmdLine << " will be Terminated after " << dwTimeOut << " milliseconds!\n";
		}

		// Close Process and Thread handles
		if (TerminateProcess(pri.hProcess, 0)) {
			// 500 ms timeout; use INFINITE for no timeout
			dwResult = WaitForSingleObject(pri.hProcess, 500);
			if (dwResult == WAIT_OBJECT_0) {
				std::cerr << "Process with Name " << strCmdLine << " is Terminated! " << endl;
				iRes = 0;
			}
			else {
			std::cerr << "Process " << GetProcessId(pri.hProcess) << "with Name " << strCmdLine << " is not Terminated!\n";
				iRes = 1;
			}
			CloseHandle(pri.hProcess);
			CloseHandle(pri.hThread);
		}
		else {
		std::cerr << "Process " << GetProcessId(pri.hProcess) << "with Name " << strCmdLine << " is not Terminated and Closed!\n";
			iRes = 1;
		}
			
	}

	return iRes;
}

int StartProcess::StartAsCmdFile(string strFile, int iTimeOut)
{
	// TODO: Добавьте сюда код реализации.
	STARTUPINFOA stInfo;
	PROCESS_INFORMATION pri;
	DWORD dwTimeOut;
	string strCmdLine;

	DWORD dwsi_size = sizeof(stInfo);
	ZeroMemory(&stInfo, dwsi_size);
	stInfo.cb = dwsi_size;
	ZeroMemory(&pri, sizeof(pri));

	if (strFile.length() == 0) {
		std::cerr << "Process Error! Can\'t Run Empty File.\n";
		return 1;
	}
	if (strCmdPath.length() == 0) {
		std::cerr << "Check Integrity Error! The Cmd.Exe Path " << strCmdPath << " does not exist!\n";
		return 1;
	}

	strCmdLine = strCmdPath + " /c \"" + strFile + "\"";

	try
	{
		dwTimeOut = iTimeOut * 1000;
	}
	catch (const std::exception&)
	{
		dwTimeOut = 0;
	}

	// Start Child Process
	if (!CreateProcessA(NULL,
		(LPSTR)strCmdLine.c_str(),
		NULL,
		NULL,
		false,
		IDLE_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS,
		NULL,
		NULL,
		&stInfo,
		&pri
	)) {
		std::cerr << "Process Creation Failed (" << GetLastError() << ").\n";
		return 1;
	}
	else
	{
		if (dwTimeOut == 0) {
			WaitForSingleObject(pri.hProcess, INFINITE);
			std::cout << "Process " << strFile << " is Terminated Successfully!\n";
		}
		else
		{
			WaitForSingleObject(pri.hProcess, dwTimeOut);
			std::cout << "Process " << strFile << " is Terminated after " << dwTimeOut << " milliseconds!\n";
		}

		// Close Process and Thread handles
		if (!TerminateProcess(pri.hProcess, 0))
			std:cerr << "Process " << GetProcessId(pri.hProcess) << " is not Terminated!\n";
		CloseHandle(pri.hProcess);
		CloseHandle(pri.hThread);
	}

	return 0;
}


int StartProcess::StartAsCScriptFiles(string strFile, int iTimeOut)
{
	// TODO: Добавьте сюда код реализации.
	STARTUPINFOA stInfo;
	PROCESS_INFORMATION pri;
	DWORD dwTimeOut;
	string strCmdLine;

	DWORD dwsi_size = sizeof(stInfo);
	ZeroMemory(&stInfo, dwsi_size);
	stInfo.cb = dwsi_size;
	ZeroMemory(&pri, sizeof(pri));

	if (strFile.length() == 0) {
		std::cerr << "Process Error! Can\'t Run Empty File.\n";
		return 1;
	}
	if (strCscriptPath.length() == 0) {
		std::cerr << "Check Integrity Error! The CScript.Exe Path " << strCscriptPath << " does not exist!\n";
		return 1;
	}

	strCmdLine = strCscriptPath + " //NoLogo \"" + strFile + "\"";

	try
	{
		dwTimeOut = iTimeOut * 1000;
	}
	catch (const std::exception&)
	{
		dwTimeOut = 0;
	}

	// Start Child Process
	if (!CreateProcessA(NULL,
		(LPSTR)strCmdLine.c_str(),
		NULL,
		NULL,
		false,
		IDLE_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS,
		NULL,
		NULL,
		&stInfo,
		&pri
	)) {
		std::cerr << "Process Creation Failed (" << GetLastError() << ").\n";
		return 1;
	}
	else
	{
		if (dwTimeOut == 0) {
			WaitForSingleObject(pri.hProcess, INFINITE);
			std::cout << "Process " << strFile << " is Terminated Successfully!\n";
		}
		else
		{
			WaitForSingleObject(pri.hProcess, dwTimeOut);
			std::cout << "Process " << strFile << " is Terminated after " << dwTimeOut << " milliseconds!\n";
		}

		// Close Process and Thread handles
		if (!TerminateProcess(pri.hProcess, 0))
			std:cerr << "Process " << GetProcessId(pri.hProcess) << " is not Terminated!\n";
		CloseHandle(pri.hProcess);
		CloseHandle(pri.hThread);
	}

	return 0;
}


int StartProcess::StartAsPoshFile(string strFile, int iTimeOut)
{
	// TODO: Добавьте сюда код реализации.
	// TODO: Добавьте сюда код реализации.
	STARTUPINFOA stInfo;
	PROCESS_INFORMATION pri;
	DWORD dwTimeOut;
	string strCmdLine;

	DWORD dwsi_size = sizeof(stInfo);
	ZeroMemory(&stInfo, dwsi_size);
	stInfo.cb = dwsi_size;
	ZeroMemory(&pri, sizeof(pri));

	if (strFile.length() == 0) {
		std::cerr << "Process Error! Can\'t Run Empty File.\n";
		return 1;
	}
	if (strWPoshPath.length() == 0) {
		std::cerr << "Check Integrity Error! The Powershel.Exe Path " << strWPoshPath << " does not exist!\n";
		return 1;
	}

	// strCmdLine = strWPoshPath + " -NoLogo -NoProfile -NoExit -ExecutionPolicy Bypass  -File \"" + strFile + "\"";
	// strCmdLine = strWPoshPath + " -NoLogo -NoProfile -WindowStyle Hidden -ExecutionPolicy Bypass  -File \"" + strFile + "\"";
	strCmdLine = strWPoshPath + " -NoLogo -NoProfile -ExecutionPolicy Bypass  -File \"" + strFile + "\"";

	try
	{
		dwTimeOut = iTimeOut * 1000;
	}
	catch (const std::exception&)
	{
		dwTimeOut = 0;
	}

	// Start Child Process
	if (!CreateProcessA(NULL,
		(LPSTR)strCmdLine.c_str(),
		NULL,
		NULL,
		false,
		IDLE_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS,
		NULL,
		NULL,
		&stInfo,
		&pri
	)) {
		std::cerr << "Process Creation Failed (" << GetLastError() << ").\n";
		return 1;
	}
	else
	{
		if (dwTimeOut == 0) {
			WaitForSingleObject(pri.hProcess, INFINITE);
			std::cout << "Process " << strFile << " is Terminated Successfully!\n";
		}
		else
		{
			WaitForSingleObject(pri.hProcess, dwTimeOut);
			std::cout << "Process " << strFile << " is Terminated after " << dwTimeOut << " milliseconds!\n";
		}

		// Close Process and Thread handles
		if (!TerminateProcess(pri.hProcess, 0))
			std:cerr << "Process " << GetProcessId(pri.hProcess) << " is not Terminated!\n";
		CloseHandle(pri.hProcess);
		CloseHandle(pri.hThread);
	}

	return 0;
}



int StartProcess::StartFile(string strFile, int iTimeOut)
{
	// TODO: Добавьте сюда код реализации.
	STARTUPINFOA stInfo;
	PROCESS_INFORMATION pri;
	string strCmdLine;

	DWORD dwsi_size = sizeof(stInfo);
	ZeroMemory(&stInfo, dwsi_size);
	stInfo.cb = dwsi_size;
	ZeroMemory(&pri, sizeof(pri));

	if (strFile.length() == 0) {
		std::cerr << "Process Error! Can\'t Run Empty File.\n";
		return 1;
	}

	strCmdLine = strCmdPath + " /c START /WAIT /REALTIME /B " + strFile;

	DWORD dwTimeOut;
	try
	{
		dwTimeOut = iTimeOut * 1000;
	}
	catch (const std::exception&)
	{
		dwTimeOut = 0;
	}

	// Start Child Process
	if (!CreateProcessA(NULL,
		(LPSTR)strCmdLine.c_str(),
		NULL,
		NULL,
		false,
		IDLE_PRIORITY_CLASS | REALTIME_PRIORITY_CLASS,
		NULL,
		NULL,
		&stInfo,
		&pri
	)) {
		std::cerr << "Process Creation Failed (" << GetLastError() << ").\n";
		return 1;
	}
	else
	{
		if (dwTimeOut == 0) {
			WaitForSingleObject(pri.hProcess, INFINITE);
			std::cout << "Process " << strFile << " is Terminated Successfully!\n";
		}
		else
		{
			WaitForSingleObject(pri.hProcess, dwTimeOut);
			std::cout << "Process " << strFile << " is Terminated after " << dwTimeOut << " milliseconds!\n";
		}

		if (!TerminateProcess(pri.hProcess, 0))
			std:cerr << "Process " << GetProcessId(pri.hProcess) << " is not Terminated!\n";
		// Close Process and Thread handles
		CloseHandle(pri.hProcess);
		CloseHandle(pri.hThread);
	}
	return 0;
}
